File: callcenter.py
from flask import Flask, request, jsonify, send_file
from flask_cors import CORS
import os
import fitz  # PyMuPDF
import openai
import boto3  # Amazon Polly 사용
from dotenv import load_dotenv
import requests
import time
import jwt

app = Flask(__name__)
CORS(app)
load_dotenv()

OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
VONAGE_APPLICATION_ID = 'e81bc289-d4dd-4d21-b213-f0181d7d2d98'
VONAGE_APPLICATION_PRIVATE_KEY_PATH = '/Users/hyeyeon/Desktop/private7.key'
NGROK_URL = "https://9bea-222-112-27-104.ngrok-free.app"
polly_client = boto3.client(
    "polly",
    region_name=os.getenv("AWS_REGION"),
    aws_access_key_id=os.getenv("AWS_ACCESS_KEY_ID"),
    aws_secret_access_key=os.getenv("AWS_SECRET_ACCESS_KEY")
)

# 전역 토큰과 만료 시간 변수 초기화
jwt_token = None
jwt_expiration = 0

def initialize_pdf_content():
    pdf_path = "./data/202009_5.이동통신단말기분실보험_약관_7.pdf"
    text = extract_text_from_pdf(pdf_path)
    global pdf_text_chunks
    pdf_text_chunks = split_text(text)

def extract_text_from_pdf(pdf_path):
    with fitz.open(pdf_path) as doc:
        return ''.join([page.get_text() for page in doc])

def split_text(text, chunk_size=4000, overlap=200):
    return [text[i:i + chunk_size] for i in range(0, len(text), chunk_size - overlap)]

def generate_answer_with_openai_requests(prompt):
    headers = {"Authorization": f"Bearer {OPENAI_API_KEY}", "Content-Type": "application/json"}
    data = {"model": "gpt-3.5-turbo", "messages": [{"role": "user", "content": prompt}], "max_tokens": 500}
    response = requests.post("https://api.openai.com/v1/chat/completions", headers=headers, json=data)
    return response.json()['choices'][0]['message']['content'].strip()

def generate_tts_file(text):
    output_dir = "audio"
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)
    
    try:
        response = polly_client.synthesize_speech(
            Text=text,
            OutputFormat="mp3",
            VoiceId="Seoyeon",
            SampleRate="16000"
        )
        
        if "AudioStream" in response:
            output_path = f"{output_dir}/output-{int(time.time() * 1000)}.mp3"
            with open(output_path, "wb") as out:
                out.write(response["AudioStream"].read())
            print("음성 파일 저장 완료:", output_path)
            return output_path
        else:
            print("Error: AudioStream이 응답에 없습니다.")
            return None
    except (BotoCoreError, ClientError) as error:
        print(f"Amazon Polly TTS 요청 중 오류 발생: {error}")
        return None

def get_jwt_token():
    global jwt_token, jwt_expiration
    current_time = int(time.time())
    
    # 토큰이 없거나 만료되었으면 새 토큰 생성
    if not jwt_token or current_time >= jwt_expiration:
        with open(VONAGE_APPLICATION_PRIVATE_KEY_PATH, 'rb') as key_file:
            private_key = key_file.read()
        
        jwt_expiration = current_time + 600  # 토큰 유효기간: 10분
        payload = {
            "application_id": VONAGE_APPLICATION_ID,
            "iat": current_time,
            "exp": jwt_expiration
        }
        
        jwt_token = jwt.encode(payload, private_key, algorithm="RS256")
        print("New JWT Token Generated:", jwt_token)
    else:
        print("Using existing JWT Token")
    
    return jwt_token

def play_audio_on_call(call_uuid, audio_filename):
    audio_url = f"{NGROK_URL}/audio/{audio_filename}"
    print("Audio URL:", audio_url)

    token = get_jwt_token()  # 토큰을 가져오거나 필요시 갱신

    url = f"https://api.nexmo.com/v1/calls/{call_uuid}/stream"
    headers = {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json"
    }
    data = {
        "stream_url": [audio_url],
        "loop": 1
    }

    # 요청 보내기
    try:
        response = requests.put(url, headers=headers, json=data)
        if response.status_code == 204:
            print("Audio is now playing on the call.")
        else:
            print(f"Failed to play audio. Status Code: {response.status_code}, Response: {response.text}")
    except requests.exceptions.RequestException as e:
        print(f"Error updating call: {e}")

@app.route('/generate_response', methods=['POST'])
def generate_response():
    question = request.json.get("text")
    call_uuid = request.json.get("call_uuid")
    relevant_content = pdf_text_chunks[0]
    input_text = f"질문: {question}\n정보: {relevant_content}"
    response_text = generate_answer_with_openai_requests(input_text)
    audio_path = generate_tts_file(response_text)
    
    if audio_path:
        audio_filename = os.path.basename(audio_path)
        play_audio_on_call(call_uuid, audio_filename)
        audio_url = f"{NGROK_URL}/audio/{audio_filename}"
        return jsonify({'response': response_text, 'audio_url': audio_url})
    else:
        return jsonify({'response': response_text, 'error': '음성 파일 생성에 실패했습니다.'}), 500

@app.route('/audio/<path:filename>', methods=['GET'])
def serve_audio(filename):
    return send_file(f"audio/{filename}", mimetype="audio/mpeg")

if __name__ == '__main__':
    initialize_pdf_content()
    app.run(host='0.0.0.0', port=5001, debug=True)

File: callcenter.js
require('dotenv').config();
const express = require('express');
const expressWs = require('express-ws');
const axios = require('axios');
const bodyParser = require('body-parser');
const cors = require('cors');
const fs = require('fs');
const path = require('path');
const speech = require('@google-cloud/speech');
const { PollyClient, SynthesizeSpeechCommand } = require('@aws-sdk/client-polly');
const { Vonage } = require('@vonage/server-sdk');  // Vonage 클래스를 사용
const WebSocket = require('ws');

const app = express();
expressWs(app);
app.use(bodyParser.json());
app.use(cors());

const PORT = 8888;

// Google Cloud STT 설정
const speechClient = new speech.SpeechClient();
let currentCallUuid = null;

// Vonage 설정
const vonage = new Vonage({
  apiKey: process.env.VONAGE_API_KEY,
  apiSecret: process.env.VONAGE_API_SECRET,
  applicationId: process.env.VONAGE_APPLICATION_ID,
  privateKey: fs.readFileSync(process.env.VONAGE_APPLICATION_PRIVATE_KEY_PATH)
});

app.post('/events', (req, res) => {
  console.log('Event received:', req.body);
  if (req.body.uuid) {
    currentCallUuid = req.body.uuid; // UUID 저장
    console.log("UUID saved:", currentCallUuid);
  }
  res.status(200).send('Event received');
});

// AWS Polly 설정
const polly = new PollyClient({
  region: process.env.AWS_REGION,
  credentials: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
  },
});

let ws;  // WebSocket 객체 전역으로 설정
let audioBuffer = Buffer.alloc(0);
let responseCache = {};

// Answer URL 엔드포인트 설정
const answerHandler = (req, res) => {
  const ncco = [
    {
      "action": "talk",
      "language": "ko-KR",
      "text": "상담을 원하시면 말씀해주세요."
    },
    {
      "action": "connect",
      "endpoint": [
        {
          "type": "websocket",
          "uri": "wss://9bea-222-112-27-104.ngrok-free.app/websocket",
          "headers": {
            "Content-Type": "audio/l16;rate=16000"
          }
        }
      ]
    }
  ];
  res.json(ncco);
};

app.get('/answer', answerHandler);
app.post('/answer', answerHandler);

// WebSocket 엔드포인트 설정
app.ws('/websocket', (wsConnection, req) => {
  ws = wsConnection;
  console.log('WebSocket 연결이 수립되었습니다.');

  ws.on('message', (message) => {
    if (!(message instanceof Buffer)) {
      console.error('Invalid message format. Expected Buffer format.');
      return;
    }

    audioBuffer = Buffer.concat([audioBuffer, message]);

    if (audioBuffer.length > 16000 * 2) {
      console.log('처리 시작: 버퍼 길이', audioBuffer.length);
      processBufferedAudio(ws, audioBuffer);
      audioBuffer = Buffer.alloc(0); // 버퍼 초기화
    }
  });

  ws.on('close', () => {
    console.log('WebSocket 연결이 종료되었습니다.');
  });

  ws.on('error', (error) => {
    console.error('WebSocket 오류 발생:', error);
  });
});

async function processBufferedAudio(ws, audioBuffer) {
  try {
    const text = await transcribeSpeechToText(audioBuffer);
    if (!text) {
      console.log('STT 변환 실패: 변환된 텍스트가 비어 있습니다.');
      return;
    }

    console.log('텍스트로 변환된 고객 음성:', text);

    let responseText;
    if (responseCache[text]) {
      responseText = responseCache[text];
      console.log('캐시된 응답 사용:', responseText);
    } else {
      responseText = await generateResponseFromPythonServer(text);
      responseCache[text] = responseText;
      console.log('LangChain에서 생성된 응답:', responseText);
    }

    const audioUrl = await synthesizeTextToSpeech(responseText);
    if (!audioUrl) {
      console.log('음성 생성 실패');
      return;
    }

    // playAudioOnCall(audioUrl);  // 오디오 URL을 통해 호출에서 재생
  } catch (error) {
    console.error('오디오 처리 중 오류 발생:', error);
  }
}

async function transcribeSpeechToText(audioBuffer) {
  const audio = { content: audioBuffer.toString('base64') };
  const config = {
    encoding: 'LINEAR16',
    sampleRateHertz: 16000,
    languageCode: 'ko-KR',
  };
  const request = { audio: audio, config: config };

  try {
    const [response] = await speechClient.recognize(request);
    if (response.results.length === 0) {
      console.log('STT 변환 결과 없음');
      return '';
    }
    return response.results.map(result => result.alternatives[0].transcript).join('\n');
  } catch (error) {
    console.error('STT 요청 중 오류 발생:', error);
    return '';
  }
}

const ngrokUrl = 'https://9bea-222-112-27-104.ngrok-free.app'
// Google TTS 대신 Amazon Polly를 사용하는 TTS 함수
async function synthesizeTextToSpeech(text) {
  const outputDir = path.join(__dirname, 'audio');
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir);
  }

  const params = {
    Text: text,
    OutputFormat: 'mp3',
    VoiceId: 'Seoyeon',
    SampleRate: '16000'
  };

  try {
    const command = new SynthesizeSpeechCommand(params);
    const { AudioStream } = await polly.send(command);
    const outputFile = path.join(outputDir, `output-${Date.now()}.mp3`);
    const writeStream = fs.createWriteStream(outputFile);

    return new Promise((resolve, reject) => {
      AudioStream.pipe(writeStream);
      writeStream.on('finish', () => {
        console.log('음성 합성 완료:', outputFile);
        const externalUrl = `${ngrokUrl}/audio/${path.basename(outputFile)}`; // 외부 접근 가능한 URL 생성
        resolve(externalUrl);
      });
      writeStream.on('error', (error) => {
        console.error('음성 합성 중 오류 발생:', error);
        reject(null);
      });
    });
  } catch (error) {
    console.error('Amazon Polly TTS 요청 중 오류 발생:', error);
    return null;
  }
}

// Python 서버에서 응답 생성
async function generateResponseFromPythonServer(inputText) {
  try {
    const response = await axios.post('http://localhost:5001/generate_response', { text: inputText,
      call_uuid: currentCallUuid 
     });
    return response.data.response;
  } catch (error) {
    console.error('LangChain 서버 오류:', error.message);
    return '응답을 생성할 수 없습니다.';
  }
}

// Vonage NCCO를 사용해 전화에서 오디오 재생


app.get('/audio/:filename', (req, res) => {
  const filePath = path.join(__dirname, 'audio', req.params.filename);
  res.setHeader('Content-Type', 'audio/mpeg');
  res.sendFile(filePath);
});

app.listen(PORT, () => {
  console.log(`Server is running on port ${PORT}`);
});
